plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts

```diff
@@ -0,0 +1,261 @@
+    options: {
+      auth: AuthService;
+      discovery: DiscoveryService;
+      logger: LoggerService;
+      cache: CacheService;
+      catalogClient?: CatalogClient; // used for testing
+    },
+  ): SlackNotificationProcessor[] {
+    const slackConfig =
+      config.getOptionalConfigArray('notifications.processors.slack') ?? [];
+    return slackConfig.map(c => {
+      const directWebhookUrl = c.getOptionalString('directWebhookUrl');
+      const channelWebhookUrl = c.getOptionalString('channelWebhookUrl');
+      const ttl = c.getOptionalNumber('cache.ttl') ?? DEFAULT_CACHE_TTL;
+      return new SlackNotificationProcessor({
+        directWebhookUrl,
+        channelWebhookUrl,
+        ttl,
+        ...options,
+      });
+    });
+  }
+
+  private constructor(options: {
+    auth: AuthService;
+    discovery: DiscoveryService;
+    logger: LoggerService;
+    directWebhookUrl?: string;
+    channelWebhookUrl?: string;
+    cache: CacheService;
+    ttl: number;
+    catalogClient?: CatalogClient;
+  }) {
+    const { auth, directWebhookUrl, channelWebhookUrl, logger, cache, ttl } =
+      options;
+    this.logger = logger;
+    this.directWebhookUrl = directWebhookUrl;
+    this.channelWebhookUrl = channelWebhookUrl;
+    this.cache = cache;
+    this.ttl = ttl;
+    this.auth = auth;
+    this.catalogClient =
+      options.catalogClient ??
+      new CatalogClient({
+        discoveryApi: options.discovery,
+      });
+  }
+  getName(): string {
+    return 'Slack';
+  }
+
+  async processOptions(
+    options: NotificationSendOptions,
+  ): Promise<NotificationSendOptions> {
+    if (!this.channelWebhookUrl || options.recipients.type !== 'entity') {
+      return options;
+    }
+
+    const entityRefs = Array.isArray(options.recipients.entityRef)
+      ? options.recipients.entityRef
+      : [options.recipients.entityRef];
+
+    const outbound: Promise<void>[] = [];
+    await Promise.all(
+      entityRefs.map(async entityRef => {
+        const compoundEntityRef = parseEntityRef(entityRef);
+        // skip users as they are sent direct messages, but allow all other entity kids
+        // to have a channel id annotation.
+        if (compoundEntityRef.kind === 'user') {
+          return;
+        }
+
+        let destination;
+        try {
+          destination = await this.getChannelFromEntity(entityRef);
+        } catch (error) {
+          this.logger.error(
+            `Failed to get Slack channel for entity: ${error.message}`,
+          );
+          return;
+        }
+
+        if (!destination) {
+          this.logger.debug(`No Slack channel found for entity: ${entityRef}`);
+          return;
+        }
+
+        const payload = JSON.stringify({
+          ...options.payload,
+          destination,
+        });
+
+        this.logger.debug(`Sending Slack channel notification: ${payload}`);
+        outbound.push(
+          sendSlackNotification(this.channelWebhookUrl!, payload).catch(
```

Comment:
does this actually need some rate limiting? a p-limit somewhere

Reply:
I think using `p-throttle` would be better idea, added it to the email processor already so there'a reference and now two use cases.